### ISSUE TITLE
Feature request: XArray backend

### ISSUE THREAD

**Issue by MatthewTurk** [yt-project/yt#1890]
[XArray](https://xarray.pydata.org/) provides a pandas-like interface to regularly gridded (with missing values) data. This has a very nice analogy to the `ds.r[]` object, when it is sliced with imaginary step sizes -- i.e., covering grids or arbitrary grids.  It would be excellent to provide either a conversion from a covering grid to an xarray or to make this a returned output from some functions.

While it is feasible to do this as-is, for instance:

```
ds = yt.load("IsolatedGalaxy/galaxy0030/galaxy0030")

d = ds.r[::512j, ::512j, ::512j]

density = d["density"]
coords = [('x', d["x"][:,0,0]), ('y', d["y"][0,:,0]), ('z', d["z"][0,0,:])]
rho = xr.DataArray(density, coords=coords)
```

This requires that the data be loaded in all-at-once.  XArray provides the ability to defer loading, either by providing something as a [dask](https://dask.pydata.org) array or by implementing a backend.

Implementing a backend seems to be the most straightforward way, as it would require subclassing `AbstractDataStore`, perhaps in the mode of the opendap backend.  This would be a very cool way to make yt more interoperable with xarray.

**Comment by jhamman** [comment]
@MatthewTurk - ran across this today. If there is continued interest in this, I think we xarray-devs would be happy to provide pointers as needed. @shoyer has been tackling some of the sticky backend issues recently and its becoming much easier to build new backends. Just keep us posted here.

**Comment by MatthewTurk** [comment]
@jhamman Absolutely.  This would be amazing. What would you suggest for how to follow along -- are these in released versions?  Thank you!

**Comment by jhamman** [comment]
The biggest change just landed this week. There is a open PR adding a new backend (https://github.com/pydata/xarray/pull/2476), that may give you a reasonable overview of the complexity involved. 

I'm not super familiar with YT and the data formats it supports but it may be worth raising an issue over on the xarray side to discuss further.

**Comment by shoyer** [comment]
I'm not entirely sure it makes sense to use Xarray's backends machinery here, which is a best fit for file formats that can be mapped to netCDF. For yt, we might just want to write a direct converter, like [what xarray does for rasterio](https://github.com/pydata/xarray/blob/4bad455a801e91b329794895afa0040c868ff128/xarray/backends/rasterio_.py).

**Comment by MatthewTurk** [comment]
@shoyer @jhamman thanks for chiming in -- I will look at rasterio.  I'm not entirely sure what the best approach would be to do this, as I think in one sense we want to have xarray data serve as a backend to yt, but it would also be quite nice to be able to supply data in the reverse direction as well, with as few copies in memory as possible.  I will look through this information and ping you when moving forward.

**Comment by jhamman** [comment]
@MatthewTurk - what we're seeing a number of projects do is to build in `to_xarray()`/`from_xarray()` methods/constructors onto their own data models. Under the hood, xarray just stores numpy or dask arrays so as long as yt can take those in, a no-copy translation should be possible.

**Comment by MatthewTurk** [comment]
@jhamman That should work, yes.  One reason I was hoping to expose a bit more of the internals -- which may turn out not to be terribly useful in the long run! -- is that we have an MPI-parallel IO system, which in principle could help with lazy loading or deferment.  In practice, though, my suspicion is that the `to_xarray` and `from_xarray` functions would likely cover enough of our use cases, so I'll have a go at that.
